### PR TITLE
Initial LinkedIn commit

### PR DESCRIPTION
**Pull request by hmrajas**

### CODE REVIEW

File: src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/internal/LinkedInAuthenticatorServiceComponent.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.osgi.service.component.ComponentContext;
+import org.wso2.carbon.identity.application.authentication.framework.ApplicationAuthenticator;
+import org.wso2.carbon.identity.authenticator.LinkedIn.LinkedInAuthenticator;
+
+import java.util.Hashtable;
+
```
Comment (Shakila) [inline src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/internal/LinkedInAuthenticatorServiceComponent.java]: One empty line is enough.

File: src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+            String json = sendRequest(
+                    LinkedInAuthenticatorConstants.LINKEDIN_USERINFO_ENDPOINT,
+                    token.getParam(LinkedInAuthenticatorConstants.ACCESS_TOKEN));
+            Map<String, Object> jsonObject = JSONUtils.parseJSON(json);
+            return jsonObject;
+        } catch (Exception e) {
+            log.error(e);
+        }
+        return new HashMap<String, Object>();
+
```
Comment (Shakila) [inline src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java]: Remove unwanted empty line.

**Comment by RKathees** [comment]
Change to new folder structure.

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/internal/LinkedInAuthenticatorServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * @scr.component name="identity.application.authenticator.LinkedIn.component" immediate="true"
+ */
+public class LinkedInAuthenticatorServiceComponent {
+
+    private static Log log = LogFactory.getLog(LinkedInAuthenticatorServiceComponent.class);
+
+    /**
+     *
+     * @param ctxt
+     */
```
Comment (RKathees) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/internal/LinkedInAuthenticatorServiceComponent.java:40]: Add proper description. This needs to be fixed in entire PR.

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+package org.wso2.carbon.identity.authenticator.LinkedIn;
```
Comment (malakasilva) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:19]: Package names make it simple case

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+    /**
+     * check weather user can process or not
+     *
+     * @param request the request
+     * @return true or false
+     */
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+        if (log.isTraceEnabled()) {
```
Comment (malakasilva) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:66]: Should this be debug logs?

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        // Check commonauth got an OIDC response
+        return request.getParameter(LinkedInAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE) != null
+                && request.getParameter(LinkedInAuthenticatorConstants.OAUTH2_PARAM_STATE) != null
+                && LinkedInAuthenticatorConstants.LINKEDIN_LOGIN_TYPE.equals(getLoginType(request));
+    }
+
+    private String getLoginType(HttpServletRequest request) {
+        String state = request.getParameter(LinkedInAuthenticatorConstants.OAUTH2_PARAM_STATE);
+        if (state != null) {
+            return state.split(",")[1];
```
Comment (malakasilva) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:78]: Need to check the length beforehand. Can throw array out of bound.

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+                response.sendRedirect(loginPage);
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Error while retrieving properties. Authenticator Properties cannot be null");
+                }
+                throw new AuthenticationFailedException(
+                        "Error while retrieving properties. Authenticator Properties cannot be null");
+            }
+        } catch (IOException e) {
+            log.error("Exception while sending to the login page", e);
```
Comment (malakasilva) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:213]: Logging and throwing both not required. Remove log.

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+                        .setClientId(clientId)
+                        .setRedirectURI(callbackurl)
+                        .setResponseType(LinkedInAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE)
+                        .setState(state).buildQueryMessage();
+                String loginPage = authzRequest.getLocationUri();
+                String domain = request.getParameter("domain");
+                if (domain != null) {
+                    loginPage = loginPage + "&fidp=" + domain;
+                }
+                if (!queryString.startsWith("&")) {
```
Comment (malakasilva) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:199]: Need to check null for query string

Comment (malakasilva) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:199]: & can go into constants file

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java (hunk 4 above)
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:19]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java (hunk 5 above)
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:66]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java (hunk 6 above)
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:78]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java (hunk 7 above)
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:213]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java (hunk 8 above)
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:199]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+    public boolean canHandle(HttpServletRequest request) {
+        if (log.isTraceEnabled()) {
+            log.trace("Inside LinkedinOAuth2Authenticator.canHandle()");
+        }
+        // Check commonauth got an OIDC response
+        return request.getParameter(LinkedInAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE) != null
+                && request.getParameter(LinkedInAuthenticatorConstants.OAUTH2_PARAM_STATE) != null
+                && LinkedInAuthenticatorConstants.LINKEDIN_LOGIN_TYPE.equals(getLoginType(request));
+    }
+
```
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:74]: Method description added https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+    /**
+     * Get Configuration Properties
+     */
+    @Override
+    public List<Property> getConfigurationProperties() {
+        List<Property> configProperties = new ArrayList<>();
+
+        Property clientId = new Property();
+        clientId.setName(LinkedInAuthenticatorConstants.CLIENT_ID);
+        clientId.setDisplayName("Client Id");
```
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:141]: Moved to constant file https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+                String queryString = LinkedInAuthenticatorConstants.QUERY_STRING;
+                authzRequest = OAuthClientRequest
+                        .authorizationLocation(authorizationEP)
+                        .setClientId(clientId)
+                        .setRedirectURI(callbackurl)
+                        .setResponseType(LinkedInAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE)
+                        .setState(state).buildQueryMessage();
+                String loginPage = authzRequest.getLocationUri();
+                String domain = request.getParameter("domain");
+                if (domain != null) {
```
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:196]: Extra method removed https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+            AuthenticatedUser authenticatedUserObj;
+            String accessToken = oAuthResponse.getParam(LinkedInAuthenticatorConstants.ACCESS_TOKEN);
+            if (accessToken != null) {
+                Map<ClaimMapping, String> claims;
+                Map<String, Object> userClaims = getUserClaims(oAuthResponse);
+                if (userClaims != null && !userClaims.isEmpty()) {
+                    authenticatedUserObj = AuthenticatedUser
+                            .createFederateAuthenticatedUserFromSubjectIdentifier(String.valueOf(userClaims.get("id")));
+                    authenticatedUserObj
+                            .setAuthenticatedSubjectIdentifier(String.valueOf(userClaims.get("lastName")));
```
Comment (hmrajas) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/LinkedIn/LinkedInAuthenticator.java:285]: Moved to constant file https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/pull/3

File: component/org.wso2.carbon.identity.authenticator.LinkedIn.connector.iml [hunk 13]
```diff
@@ -0,0 +1,1642 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (hmrajas) [inline component/org.wso2.carbon.identity.authenticator.LinkedIn.connector.iml:1]: Removed https://github.com/wso2-extensions/identity-outbound-auth-linkedIn/commit/76b65d54ee430bbe9aab737141878736c0e47439

**Pull request merged**